staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,414 @@
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+	// `reason` is a unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	// `message` is a human-readable message indicating details about last transition.
+	Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+}
+
+// FlowSchemaConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `spec` is the specification of the desired behavior of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// `status` is the current status of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+// +union
+type PriorityLevelConfigurationSpec struct {
+	// `queuingType` indicates whether this priority level does
+	// queuing or is exempt.  Valid values are "Queuing" and "Exempt",
+	// and this field defaults to "Queuing".  "Exempt" means that
+	// requests of this priority level are not subject to concurrency
+	// limits (and thus are never queued) and do not detract from the
+	// concurrency available for non-exempt requests.
+	// Required.
+	// +unionDiscriminator
+	QueuingType PriorityLevelQueueingType `json:"queuingType" protobuf:"varint,1,opt,name=queuingType"`
```

Comment:
Just "Type" is a sufficient name?

Reply:
I like being more specific because other sorts of typing may be added in the future and we do not want this union discriminator to become a garbage can amalgalmation of all typing info.

Reply:
Type is the type of the union. I don't think we'll stuff another union into this struct?

Reply:
My concern is what happens when we put a union somewhere else below this level of structure, for example because we evolve some aspect of queuing to be configured with a union.  Not all of queuing, just some aspect of it.  That is, a field of QueuingConfiguration holds a union.  That is in some sense a typing of the whole PriorityLevelConfiguration, but that distinction should not be folded into one grand top-level "type" field.

Reply:
But those would be in a different scope and not confusing? E.g. imagine we add multiple ways to specify the shares, then we might have something like this:

```yaml
type: Queueing
queueing:
  type: ExactConcurrencyLimit
  concurrencyLimit: 3
```

I think that's OK?

Reply:
Let me first make sure I understand your example.  You are supposing that in the future we might introduce the option to specify AssuredConcurrencyValue (ACV) directly instead of specifying AssuredConcurrencyShares (ACS), and this does not change the fact that a priority level continues to have Queues, HandSize, and QueueLengthLimit.  I see in https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#go-tags the remark that "All the fields must be optional (beside the discriminator) and will be inlucded as members of the union" --- in other words, the union can not be flattened into a larger structure.  So a v1alpha2 instance would look more like

```
type: Queuing
queuing:
  concurrency:
    type: Absolute
    value: 3
  queues: 30
  handSize: 5
  queueLengthLimit: 10
```

--- right?

That (either as you wrote it or as I wrote it) would be much better than splitting the one Queuing type into QueuingAbsolute and QueuingByShares.

My remaining concern is that it looks a little disturbing to have two fields named "type".  I think it might be clearer to give the two "type" fields more specific names.  However, if the way I wrote it is right, the two "type" fields are separated by two levels of structure and narrowness of the context of the inner "type" field is pretty clear; I could live with that.

Reply:
i think a short name `type` is better as a union discriminator.